### ISSUE TITLE
Possible to dump all available readout elements?

### ISSUE THREAD

**Issue by alexandertuna** [AIDASoft/DD4hep#1297]
Hi,

@headunderheels and I are simulating an electromagnetic calorimeter with dd4hep and a compact detector description. Thanks for the very nice code! We have a small question: is it possible to dump *all* readout elements somehow?

For example, if we simulate a particle passing through the ecal, we can get a collection of readout elements where the particle deposited energy. But is it possible to get a list of *all* readout elements, without simulating a particle? It would be useful for us to have this info.

Thanks for your help, we're new to this!

---

Our ecal, for example: https://github.com/madbaron/detector-simulation/blob/KITP_10TeV/geometries/MuColl_10TeV_v0A/ECalEndcap_o2_v01_02.xml

```
<lccdd>

	<readouts>
		<readout name="ECalEndcapCollection">
			<segmentation type="CartesianGridXY" grid_size_x="ECal_cell_size" grid_size_y="ECal_cell_size" />
			<id>${GlobalCalorimeterReadoutID}</id>
		</readout>
	</readouts>

	<detectors>
		<detector name="ECalEndcap" type="GenericCalEndcap_o1_v01" id="DetID_ECal_Endcap" readout="ECalEndcapCollection" vis="ECALVis" >

        and so on
```

**Comment by atolosadelgado** [comment]
Hi

The following snippet shows how to loop over the detector element map and  sensitive detector map:

```cpp
#include "DD4hep/Detector.h"
#include "DD4hep/DD4hepUnits.h"
#include "DD4hep/BitFieldCoder.h"

R__LOAD_LIBRARY(libDDCore) 

auto lcdd = &(dd4hep::Detector::getInstance());
lcdd->fromCompact( path_to_main_XML )

// map of detectors, geometry
for( auto & [name, det] : lcdd->detectors() ) std::cout << name.c_str() << std::endl;

// map of sensitive detectors, readout
for( auto & [name, sd] : lcdd->sensitiveDetectors() ) std::cout << name.c_str() << std::endl;

```

Best,
Alvaro

**Comment by alexandertuna** [comment]
Thanks for the fast response @atolosadelgado ! We’ll try this asap and let you know

**Comment by alexandertuna** [comment]
Hi again @atolosadelgado , I think this is almost what we want. We're interested in dumping a full list of readout *elements*, not just readouts. In our example, we have:

`<detector>` -> `<readout>` -> `<segmentation>` -> `CartesianGridXY`

We're interested in dumping each element of the `CartesianGridXY`. I'll look for methods within dd4hep for finding this, but feel free to post any suggestions if you have them.

**Comment by vvolkl** [comment]
Yes, this is not that straightforward, as it depends on the segmentation. Previous discussion here: https://github.com/AIDASoft/DD4hep/issues/580

**Comment by atolosadelgado** [comment]
> Hi again @atolosadelgado , I think this is almost what we want. We're interested in dumping a full list of readout _elements_, not just readouts. In our example, we have:
> 
> `<detector>` -> `<readout>` -> `<segmentation>` -> `CartesianGridXY`
> 
> We're interested in dumping each element of the `CartesianGridXY`. I'll look for methods within dd4hep for finding this, but feel free to post any suggestions if you have them.

Hi,

the last line of the snippet provides what you are looking for :)

```cpp
#include "DD4hep/Detector.h"
#include "DD4hep/DD4hepUnits.h"
#include "DD4hep/BitFieldCoder.h"

R__LOAD_LIBRARY(libDDCore) 

auto lcdd = &(dd4hep::Detector::getInstance());
lcdd->fromCompact( path_to_main_XML )

// map of detectors, geometry
for( auto & [name, det] : lcdd->detectors() ) std::cout << name.c_str() << std::endl;

// map of sensitive detectors, readout
for( auto & [name, sd] : lcdd->sensitiveDetectors() ) std::cout << name.c_str() << std::endl;

// map of readouts
for( auto & [name, ro_b] : lcdd->readouts() ){ dd4hep::Readout r = ro_b; std::cout << r.segmentation().type()<< std::endl;}

```

**Comment by BrieucF** [comment]
Hi @alexandertuna , 

It looks indeed that, since the segmentation does not know, a priori, what is the extent of the volume to be segmented there is no magic function to get what you want. I think you have to retrieve the segmentation and the dimension of the volume at the same time to be able to extract extremas. 

See for instance: https://github.com/key4hep/k4geo/blob/main/detectorCommon/src/DetUtils_k4geo.cpp#L382 .Which is used (for a different segmentation) e.g. here: https://github.com/HEP-FCC/k4RecCalorimeter/blob/main/RecFCChhCalorimeter/src/components/CreateFCChhCaloNeighbours.cpp#L92 

Cheers,
Brieuc

**Comment by alexandertuna** [comment]
Thanks everyone. I believe @vvolkl and @BrieucF understand my request - it is the same as https://github.com/AIDASoft/DD4hep/issues/580 , if I understand correctly.

Thanks for these links, @BrieucF . It seems there is no magic way to list all the readout channels without e.g. calculating `cellsEta` and `minEtaID` with `ceil/floor`. That's good to know. We'll need to think about if we want to repeat this logic ourselves.

**Comment by MarkusFrankATcernch** [comment]
As you have already noted issue https://github.com/AIDASoft/DD4hep/issues/580 already states that that enumerating all sensitive elements within a sensitive volume in general is impossible. One can enumerate if the volume is of a very concrete shape (such as a box, a cylinder, etc.), but otherwise is very time consuming and not possible in a only half way straight forward way. At the end one would always to sort of scan through an approximation and for each cell check if the edges are "inside" or "outside" the concrete sensitive volume. This is very expensive.

If you can come up with a sort of generic strategy, please let me know! I am happy to implement.

**Comment by andresailer** [comment]
If someone comes up with a generic solution, we can revisit this issue.
